<launch>

  <arg name="rviz" default="true" />
  <arg name="scanPeriod" default="0.2" />
  <node pkg="tf" type="static_transform_publisher" name="camera_init_to_map" args="0 0 0 1.570795 0 1.570795 /map /camera_init 10" />
  <node pkg="tf" type="static_transform_publisher" name="base_link_to_camera" args="0 0 0 -1.570795 -1.570795 0 /camera /base_link 10" />
  <node pkg="loam_velodyne" type="imageProjection" name="imageProjection" output="screen">
    <param name="lidar" value="VLP-16" /> <!-- options: VLP-16  HDL-32  HDL-64E -->
    <param name="scanPeriod" value="$(arg scanPeriod)" />
    <remap from="/lslidar_point_cloud" to="/lslidar_point_cloud" />
  </node>

  <node pkg="loam_velodyne" type="featureAssociation" name="featureAssociation" output="screen" respawn="true">
    <param name="scanPeriod" value="$(arg scanPeriod)" />
  </node>
   <node pkg="loam_velodyne" type="mapOptmization" name="mapOptmization" output="screen">
  </node>
  <node pkg="loam_velodyne" type="transformFusion" name="transformFusion" output="screen">
  </node>
   <node pkg="loam_velodyne" type="laserImage_node" name="laserImage_node" output="screen">
    <!-- 1 is right,2 is back, 3 is up,4 is left-->
    <param name="camera_number" value="4" />
    <param name="intrinsic" value="$(find loam_velodyne)/config/fisheye.ini" />
    <param name="extrinsic1" value="$(find loam_velodyne)/config/extrinsic1.yaml" />
    <param name="extrinsic2" value="$(find loam_velodyne)/config/extrinsic2.yaml" />
    <param name="extrinsic3" value="$(find loam_velodyne)/config/extrinsic3.yaml" />
    <param name="extrinsic4" value="$(find loam_velodyne)/config/extrinsic4.yaml" />
    <param name="camera1_pre_theta" value="180" />
    <param name="camera2_pre_theta" value="90" />
    <param name="camera3_pre_theta" value="0" />
    <param name="camera4_pre_theta" value="-90" />
    <!--resize image-->
    <param name="resizeImageTimes" value="4" />
  </node>
  <node pkg="loam_velodyne" type="backendOptim" name="backendOptim" output="screen">
    <!-- if calMethod is distribution,then the calculating pose is
    distribution method,if calmethod is center,the calculating
    method is center-->
    <param name="carId" value="master" />
    <param name="calMethod" value="center" />
    <!-- send params 14-->
    <param name="pos_quantization_bits" value="10" />
    <!-- keyframe registration params -->
    <param name="max_keyframes_per_update" value="10" />
    <param name="keyframe_delta_trans" value="2" />
    <param name="keyframe_delta_angle" value="2" />
    <param name="keyframe_delta_time" value="1" />
    <!-- loop closure params -->
    <param name="distance_thresh" value="5.0" />
    <param name="child_master_distance_thresh" value="5.0" />
    <param name="accum_distance_thresh" value="8.0" />
    <param name="min_edge_interval" value="5.0" />
    <!--- 0.5-->
    <param name="fitness_score_thresh" value="3" />
    <param name="childmaster_fitness_score_thresh" value="8" />
    <param name="search_radius_num" value="25" />
    <!-- cloud matching params -->
    <param name="registration_method" value="NDT_OMP" />
    <param name="ndt_resolution" value="0.5" />
    <param name="ndt_num_threads" value="0" />
    <param name="ndt_nn_search_method" value="DIRECT7" />
    <!-- information matrix params -->
    <param name="gps_edge_stddev" value="1000.0" />
    <param name="floor_edge_stddev" value="20.0" />
    <!-- update params -->
    <param name="graph_update_interval" value="3.0" />
    <param name="map_cloud_update_interval" value="10.0" />
    <!--compress cloud -->
    <param name="compressCloud" value="low" />
  </node>

  <group if="$(arg rviz)">
    <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find loam_velodyne)/rviz_cfg/loam_loop.rviz" />
  </group>

</launch>
